python/ml4ir/base/io/file_io.py

```diff
@@ -1,298 +1,181 @@
-
-    Args:
-        infile: path to the csv input file; can be hdfs path
-        sep: separator to use for loading file
-        index_col: column to be used as index
-        log_path: boolean specifying whether path should be logged
-        log: logging object
-
-    Returns:
-        pandas dataframe
-    """
-    if log_path:
-        log.info("Loading dataframe from path : {}".format(infile))
-
-    if infile.startswith(HDFS_PREFIX):
-        # NOTE: Move to fully spark dataframe based operations in the future
-        return spark_io.read_df(infile).toPandas()
-    elif infile.endswith(".gz"):
-        fp = gzip.open(os.path.expanduser(infile), "rb")
-    else:
-        fp = open(os.path.expanduser(infile), "r")
-
-    # Redirect stderr to custom string IO
-    stderr_old = sys.stderr
-    sys.stderr = bad_lines_io = StringIO()
-
-    try:
-        df: pd.DataFrame = pd.read_csv(
-            fp,
-            sep=sep,
-            index_col=index_col,
-            skipinitialspace=True,
-            quotechar='"',
-            escapechar="\\",
-            error_bad_lines=False,
-            warn_bad_lines=True,
-            engine="c",
-        )
-    except Exception as e:
-        if log:
-            log.info("Error while reading : {}\n{}".format(fp, e))
-        return None
-
-    # Get the bad line string value and close the string IO
-    bad_lines = bad_lines_io.getvalue()
-    bad_lines_io.close()
-    sys.stderr = stderr_old
-
-    # Log any bad lines
-    if bad_lines and log:
-        log.info("Bad lines were found in the file : {}\n{}".format(infile, bad_lines))
-
-    fp.close()
-    return df
-
-
-def read_df_list(infiles, sep=",", index_col=None, log_path=True, log=None) -> pd.DataFrame:
-    """
-    Load a pandas dataframe from a list of files
-
-    Args:
-        infiles: paths to the csv input files; can be hdfs paths
-        sep: separator to use for loading file
-        index_col: column to be used as index
-        log_path: boolean specifying whether path should be logged
-        log: logging object
-
-    Returns:
-        pandas dataframe
-    """
-    if log:
-        log.info("Reading {} files from [{}, ..".format(len(infiles), infiles[0]))
-    return pd.concat(
-        [
-            read_df(infile, sep=sep, index_col=index_col, log_path=False, log=log)
-            for infile in infiles
-        ]
-    )
-
-
-def write_df(df, outfile: str = None, sep: str = ",", index: bool = True, log=None) -> str:
-    """
-    Write a pandas dataframe to a file
-
-    Args:
-        df: dataframe to be written
-        outfile: path to the csv output file; can NOT be hdfs path currently
-        sep: separator to use for loading file
-        index: boolean specifying if index should be saved
-        log: logging object
-
-    Returns:
-        dataframe in csv form if outfile is None
-
-    NOTE: Does not support spark write
-    """
-    if outfile and outfile.startswith(HDFS_PREFIX):
+
+class FileIO(object):
```

Comment:
I like this! Defining FileIO and then having the LocalIO and SparkIO, is clean, slick etc. nice!